rust/sedona-datasource/src/format.rs

```diff
@@ -0,0 +1,717 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{any::Any, collections::HashMap, fmt::Debug, sync::Arc};
+
+use arrow_schema::{Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::{
+    config::ConfigOptions,
+    datasource::{
+        file_format::{file_compression_type::FileCompressionType, FileFormat, FileFormatFactory},
+        listing::PartitionedFile,
+        physical_plan::{
+            FileGroupPartitioner, FileMeta, FileOpenFuture, FileOpener, FileScanConfig,
+            FileSinkConfig, FileSource,
+        },
+    },
+};
+use datafusion_catalog::{memory::DataSourceExec, Session};
+use datafusion_common::{not_impl_err, DataFusionError, GetExt, Result, Statistics};
+use datafusion_physical_expr::{LexOrdering, LexRequirement, PhysicalExpr};
+use datafusion_physical_plan::{
+    filter_pushdown::{FilterPushdownPropagation, PushedDown},
+    metrics::ExecutionPlanMetricsSet,
+    ExecutionPlan,
+};
+use futures::{StreamExt, TryStreamExt};
+use object_store::{ObjectMeta, ObjectStore};
+use sedona_common::sedona_internal_err;
+
+use crate::spec::{ExternalFormatSpec, Object, OpenReaderArgs, SupportsRepartition};
+
+/// Create a [FileFormatFactory] from a [ExternalFormatSpec]
+///
+/// The FileFormatFactory is the object that may be registered with a
+/// SessionStateBuilder to allow SQL queries to access this format.
+#[derive(Debug)]
+pub struct ExternalFormatFactory {
+    spec: Arc<dyn ExternalFormatSpec>,
+}
+
+impl ExternalFormatFactory {
+    pub fn new(spec: Arc<dyn ExternalFormatSpec>) -> Self {
+        Self { spec }
+    }
+}
+
+impl FileFormatFactory for ExternalFormatFactory {
+    fn create(
+        &self,
+        _state: &dyn Session,
+        format_options: &HashMap<String, String>,
+    ) -> Result<Arc<dyn FileFormat>> {
+        Ok(Arc::new(ExternalFileFormat {
+            spec: self.spec.with_options(format_options)?,
+        }))
+    }
+
+    fn default(&self) -> Arc<dyn FileFormat> {
+        Arc::new(ExternalFileFormat {
+            spec: self.spec.clone(),
+        })
+    }
+
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+}
+
+impl GetExt for ExternalFormatFactory {
+    fn get_ext(&self) -> String {
+        self.spec.extension().to_string()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ExternalFileFormat {
+    spec: Arc<dyn ExternalFormatSpec>,
+}
+
+impl ExternalFileFormat {
+    pub fn new(spec: Arc<dyn ExternalFormatSpec>) -> Self {
+        Self { spec }
+    }
+}
+
+#[async_trait]
+impl FileFormat for ExternalFileFormat {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn get_ext(&self) -> String {
+        self.spec.extension().to_string()
+    }
+
+    fn get_ext_with_compression(
+        &self,
+        _file_compression_type: &FileCompressionType,
+    ) -> Result<String> {
+        not_impl_err!("extension with compression type")
```

Comment:
```suggestion
        match _file_compression_type {
            FileCompressionType::UNCOMPRESSED => Ok(self.get_ext()),
            _ => not_impl_err!("extension with compression type"),
        }
```
a minor improvement for the uncompressed case.